### PR TITLE
[INTEL MKL] Enabled MIN_FIRST support and primitive caching for MKL-DNN Quantize OP

### PR DESCRIPTION
**Pull request by rgomathi**
This PR will enable
1. Min first mode support in the MKL-DNN quantize Op and
2. Primitive Caching for the same Op

Code changes involve
1. Rewriting the rules in the mkl_layout_pass.cc to allow mkl-dnn quantize function being called for MIN_FIRST mode as well
2. Appropriate code added in mkl_quantize_op.cc to support MIN_FIRST mode and primitive caching
3. Added 2 more tests in mkl_quantize_op_test.cc specific to MIN_FIRST mode.

### CODE REVIEW

**Comment by gbaned** [comment]
@rgomathi Can you please resolve conflicts? Thanks!

File: tensorflow/core/kernels/mkl_quantize_op_test.cc [hunk 1]
```diff
@@ -95,4 +95,95 @@ TEST_F(MklQuantizeV2OpTest, small_int8) {
...
+  Tensor expected(allocator(), DT_QUINT8, TensorShape({8}));
+  test::FillValues<quint8>(&expected, {32, 64, 96, 128, 159, 191, 223, 255});
+  test::ExpectTensorEqual<quint8>(expected, *GetOutput(0));
+  const float output_min = GetOutput(1)->flat<float>()(0);
+  const float output_max = GetOutput(2)->flat<float>()(0);
+  EXPECT_NEAR(0.0f, output_min, 1e-5f);
+  EXPECT_NEAR(0.8f, output_max, 1e-5f);
+}
+
+TEST_F(MklQuantizeV2OpTest, small_minfirst_int) {
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_quantize_op_test.cc]: Just curious. The testcase name is `small_minfirst_int`. Where is `int` used?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR! I appreciate the detailed PR description, and I'm sorry for my delay!

File: tensorflow/core/graph/mkl_layout_pass.cc [hunk 2]
```diff
@@ -2420,6 +2424,24 @@ void MklLayoutRewritePass::CopyAttrsConvCheckConstFilter(const Node* orig_node,
   CopyFormatAttrsConv(orig_node, nb, strides, dilations, change_format);
 }
 
+void MklLayoutRewritePass::CopyAttrsQuantizeV2(const Node* orig_node,
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/graph/mkl_layout_pass.cc]: with rebase to master i could use CopyAttrsAll itself..

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 3]
```diff
@@ -49,6 +49,196 @@ namespace tensorflow {
...
+                               memory::desc dst_md)
+      : src_dims(src_dims), src_md(src_md), dst_md(dst_md) {}
+};
+
+class MklReorderWithScalePrimitive : public MklPrimitive {
+ public:
+  explicit MklReorderWithScalePrimitive(
+      const memory* from, const memory* to,
+      const MklReorderWithScaleFwdParams& fwdParams) {
+    // create reorder primitive
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Done. Taken care in other places as well

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 4]
```diff
@@ -49,6 +49,196 @@ namespace tensorflow {
...
+  // set data handles
+  void SetMemory(const memory* from, const memory* to) {
+    context_.src_mem->set_data_handle(from->get_data_handle());
+    context_.dst_mem->set_data_handle(to->get_data_handle());
+  }
+
+ private:
+  // Primitive reuse context for reorder
+  struct ReorderContext {
+    // MKLDNN memory
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Done.

**Comment by rgomathi** [comment]
Hi @penpornk , Thanks for the comments, will resolve them and get back to you ASAP

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 5]
```diff
@@ -49,6 +49,196 @@ namespace tensorflow {
...
+    auto const& post_op_params = fwdParams.post_op_params;
+    mkldnn::primitive_attr post_ops_attr;
+
+    if (post_op_params.name == "scale") {
+      DCHECK_EQ(post_op_params.param.size(), 1);
+      std::vector<float> scales;
+      scales.push_back(post_op_params.param[0]);
+      post_ops_attr.set_output_scales(0, scales);
+    } else {
+      DCHECK(post_op_params.name == "scale");
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: My bad, fixed the code now...

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 6]
```diff
@@ -49,6 +49,196 @@ namespace tensorflow {
...
+  }
+
+  void SetReorder(const memory* from, const memory* to, MklPrimitive* op,
+                  const MklReorderWithScaleFwdParams& fwdParams) {
+    string key = CreateKey(from, to, fwdParams);
+    this->SetOp(key, op);
+  }
+};
+
+/// Fuction to find(or create) a reorder from memory pointed by
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Understood and made changes in other places as well.

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 7]
```diff
@@ -49,6 +49,196 @@ namespace tensorflow {
...
+/// from to memory pointed by to, it will create primitive or
+/// get primitive from pool if it is cached.
+/// Returns the primitive.
+template <typename T>
+inline primitive FindOrCreateReorder(
+    const memory* from, const memory* to,
+    const MklReorderWithScaleFwdParams& fwdParams) {
+  CHECK_NOTNULL(from);
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Done..

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 8]
```diff
@@ -57,19 +247,42 @@ class MklQuantizeV2Op : public OpKernel {
...
+                                          "only supported for mode 'SCALED', "
+                                          "but mode is '" +
+                                          mode_string + "'."));
+      round_mode_ = ROUND_HALF_TO_EVEN;
+    }
   }
 
-  ~MklQuantizeV2Op() {}
+  ~MklQuantizeV2Op() {
+    if (this->minfirst_input_ != nullptr) {
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Removed..

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 9]
```diff
@@ -57,19 +247,42 @@ class MklQuantizeV2Op : public OpKernel {
...
+      delete this->minfirst_input_;
+      minfirst_input_ = nullptr;
+    }
+  }
+
+  float* GetMinfirstInputBuf(int size) {
+    if (!minfirst_input_) {
+      minfirst_input_ = new float[size];
+    }
+    return minfirst_input_;
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Changed the code accordingly.

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 10]
```diff
@@ -135,7 +348,20 @@ class MklQuantizeV2Op : public OpKernel {
...
             ? src_mkl_shape.GetMklLayout()
             : memory::desc(src_dims, MklDnnType<float>(), dst_layout_type);
-    src.SetUsrMem(src_md, &src_tensor);
+
+    // If the mode is min_first, input data has to be subtracted from
+    // min_range, before being scaled
+    auto flat_input = input.flat<float>().data();
+    if (mode_ == QUANTIZE_MODE_MIN_FIRST) {
+      float* minfirst_input = GetMinfirstInputBuf(input.NumElements());
+#pragma omp parallel for schedule(static)
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Rewritten the code to use parallelFor().

File: tensorflow/core/kernels/mkl_quantize_op_test.cc (hunk 1 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op_test.cc]: To differentiate the input values given to the operator, used _int to note them as integers (includes -ve values). Next test has all unsigned inputs, so used _uint. Hope it makes sense.

**Comment by rgomathi** [comment]
Hi @penpornk , I'm extremely Sorry for the delay. Made the changes recommended. Please review and let me know your comments. Thanks.

**Comment by gbaned** [comment]
@rgomathi Can you please resolve conflicts? Thanks!

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 11]
```diff
@@ -57,24 +244,100 @@ class MklQuantizeV2Op : public OpKernel {
...
+    AllocateOutputSetMklShape(ctx, 1, &output_min_tensor, min_tf_shape,
+                              min_mkl_shape);
+    TensorShape max_tf_shape = {};
+    MklDnnShape max_mkl_shape;
+    max_mkl_shape.SetMklTensor(false);
+    Tensor* output_max_tensor = nullptr;
+    AllocateOutputSetMklShape(ctx, 2, &output_max_tensor, max_tf_shape,
+                              max_mkl_shape);
+
+    // Estimate scale for qunatization
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_quantize_op.cc]:
```suggestion
    // Estimate scale for quantization.
```

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 12]
```diff
@@ -57,24 +244,100 @@ class MklQuantizeV2Op : public OpKernel {
...
+  }
 
   void Compute(OpKernelContext* ctx) override {
+    const Tensor& input = ctx->input(0);
     const float input_min_range = ctx->input(1).flat<float>()(0);
     const float input_max_range = ctx->input(2).flat<float>()(0);
     float min_range = std::min(0.0f, input_min_range);
     float max_range;
-    OP_REQUIRES(ctx, (input_max_range > input_min_range),
+    OP_REQUIRES(ctx, !(input_max_range < input_min_range),
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_quantize_op.cc]: Nit: I think the original expression is more straightforward and easier to understand.
```suggestion
    OP_REQUIRES(ctx, input_max_range > input_min_range,
```

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 13]
```diff
@@ -135,7 +401,29 @@ class MklQuantizeV2Op : public OpKernel {
...
+    if (mode_ == QUANTIZE_MODE_MIN_FIRST) {
+      float* minfirst_input = GetMinfirstInputBuf(input.NumElements());
+      const Eigen::TensorOpCost cost(
+          sizeof(float), /*load bytes*/
+          sizeof(float), /*saved bytes*/
+          /*sub cost*/ Eigen::TensorOpCost::AddCost<float>());
+
+      const CPUDevice& d = ctx->eigen_device<CPUDevice>();
+      auto ParallelSub = [&](int64 start, int64 end) {
+        for (int i = start; i < end; i++) {
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_quantize_op.cc]: Nit:
```suggestion
        for (int i = start; i < end; ++i) {
```

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 14]
```diff
@@ -135,7 +401,29 @@ class MklQuantizeV2Op : public OpKernel {
...
+
+    // If the mode is min_first, input data has to be subtracted from
+    // min_range, before being scaled
+    auto flat_input = input.flat<float>().data();
+    if (mode_ == QUANTIZE_MODE_MIN_FIRST) {
+      float* minfirst_input = GetMinfirstInputBuf(input.NumElements());
+      const Eigen::TensorOpCost cost(
+          sizeof(float), /*load bytes*/
+          sizeof(float), /*saved bytes*/
+          /*sub cost*/ Eigen::TensorOpCost::AddCost<float>());
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_quantize_op.cc]: Nit: Can we make the comment placement consistent?
```suggestion
          Eigen::TensorOpCost::AddCost<float>() /*sub cost*/);
```

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 15]
```diff
@@ -57,24 +244,100 @@ class MklQuantizeV2Op : public OpKernel {
...
+
+    out_data[0] = (src_data[0] - min_range) * scale_factor;
+    output_min_tensor->flat<float>()(0) = min_range;
+    output_max_tensor->flat<float>()(0) = max_range;
+
+    return;
+  }
 
   void Compute(OpKernelContext* ctx) override {
+    const Tensor& input = ctx->input(0);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_quantize_op.cc]: Maybe move `src_index` declaration to above this line and use it?
```suggestion
    const Tensor& input = ctx->input(src_idx);
```

File: tensorflow/core/kernels/mkl_quantize_op_test.cc (hunk 1 above)
Comment (penpornk) [reply tensorflow/core/kernels/mkl_quantize_op_test.cc]: Got it. Thank you for the explanation!

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank for the fixes and sorry for my delay! I have some more minor formatting comments and we should be good to go.

File: tensorflow/core/graph/mkl_layout_pass.cc [hunk 16]
```diff
@@ -1608,6 +1610,15 @@ rinfo_.push_back({csinfo_.tanh_grad,
 
       return false;
     }
+    if (mode_string == "MIN_FIRST") {
+      if (type != DT_QUINT8) {
+        VLOG(1) << "QuantizeOpRewrite: For MIN_FIRST mode the data type is "
+                << "not DT_UINT8."
+                << "This case is not optimized by Intel MKL, "
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/graph/mkl_layout_pass.cc]: Done..

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 17]
```diff
@@ -268,26 +265,79 @@ class MklQuantizeV2Op : public OpKernel {
...
+      delete minfirst_input_;
       minfirst_input_ = nullptr;
     }
   }
 
   float* GetMinfirstInputBuf(int size) {
     if (!minfirst_input_) {
       minfirst_input_ = new float[size];
+      minfirst_input_size_ = size;
+    } else if (size != minfirst_input_size_) {
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Thanks for the suggestion, changed the code accordingly.

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 18]
```diff
@@ -57,24 +244,100 @@ class MklQuantizeV2Op : public OpKernel {
...
+      delete minfirst_input_;
+      minfirst_input_ = new float[size];
+      minfirst_input_size_ = size;
+    }
+
+    return minfirst_input_;
+  }
+
+  void Compute_Scalar(OpKernelContext* ctx, float min_range, float max_range) {
+    // TO-DO - Scalar support has to be added for SCALE mode
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Done!

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 19]
```diff
@@ -57,24 +244,100 @@ class MklQuantizeV2Op : public OpKernel {
...
+    } else if (size != minfirst_input_size_) {
+      delete minfirst_input_;
+      minfirst_input_ = new float[size];
+      minfirst_input_size_ = size;
+    }
+
+    return minfirst_input_;
+  }
+
+  void Compute_Scalar(OpKernelContext* ctx, float min_range, float max_range) {
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Done

File: tensorflow/core/kernels/mkl_quantize_op.cc (hunk 11 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Done...

File: tensorflow/core/kernels/mkl_quantize_op.cc (hunk 12 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: I changed this to fix a unit test where min and max ranges were same. In such scenarios, we need not send error message. Also I referred the code in quantize_op.cc which fixes this in the same way. Hope this make sense.

File: tensorflow/core/kernels/mkl_quantize_op.cc (hunk 13 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Done..

File: tensorflow/core/kernels/mkl_quantize_op.cc (hunk 14 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Done.

File: tensorflow/core/kernels/mkl_quantize_op.cc (hunk 15 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Done

**Comment by rgomathi** [comment]
Hi @penpornk , Thanks for the comments and I changed the code accordingly. Please review and let me know. Thanks

File: tensorflow/core/kernels/mkl_quantize_op.cc (hunk 12 above)
Comment (penpornk) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Got it. Would this work? It's equivalent and more straightforward.
```suggestion
    OP_REQUIRES(ctx, input_max_range >= input_min_range,
```

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 20]
```diff
@@ -174,7 +430,29 @@ class MklQuantizeV2Op : public OpKernel {
...
+
+    // If the mode is min_first, input data has to be subtracted from
+    // min_range, before being scaled
+    auto flat_input = input.flat<float>().data();
+    if (mode_ == QUANTIZE_MODE_MIN_FIRST) {
+      float* minfirst_input = GetMinfirstInputBuf(input.NumElements());
+      const Eigen::TensorOpCost cost(
+          sizeof(float), /*load bytes*/
+          sizeof(float), /*saved bytes*/
+          Eigen::TensorOpCost::AddCost<float>()/*sub cost */);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_quantize_op.cc]: Nit:
```suggestion
          Eigen::TensorOpCost::AddCost<float>() /*sub cost*/);
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the fixes and I'm so sorry for my delay! I have two more minor comments.

**Comment by rgomathi** [comment]
Fixed them....

**Review by penpornk — APPROVED** [review]
Thank you for your patience!

**Comment by gbaned** [comment]
@rgomathi Could you please address Ubuntu Sanity errors? Thanks!

**Comment by gbaned** [comment]
@rgomathi Never mind. Ubuntu Sanity build successful after re-trigger tests. Thanks!

**Comment by nammbash** [comment]
Thank you @penpornk

**Pull request merged**